RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,890 @@
+
+Various attempts have been made to attempt to get the benefits of applicatives within the existing syntax, but most end up involving confusing boilerplate, and make it easy to provide arguments in the wrong order because they cannot be named (in contrast to `let! ... and! ... return ...` which forces each argument to be named right next to its value in order to be used inside the `return`). It tends to be the case that even the authors of these experiments consider them abuses of the existing language features and recommend against them.
+
+<details>
+  <summary>Expand: An example of trying to simulate applicatives using the existing CE syntax</summary>
+  <p>
+
+```fsharp
+type 'a Foo = private Foo of 'a
+
+[<RequireQualifiedAccess>]
+module Foo =
+
+    let ofValue (a : 'a) : 'a Foo = Foo a
+
+    let apply ((Foo f) : ('a -> 'b) Foo) ((Foo a) : 'a Foo) : 'b Foo =
+        Foo (f a)
+
+
+type FooBuilder () =
+
+    member __.Yield (_ : unit) =
+        id
+
+    [<CustomOperation("apply")>]
+    member __.Apply (f : 'a Foo -> ('b -> 'c) Foo, foo : 'b Foo) : 'a Foo -> 'c Foo =
+        f >> (fun ff -> Foo.apply ff foo)
+
+    [<CustomOperation("into")>]
+    member __.Into (f : 'a Foo -> 'b Foo, a : 'a) : 'b Foo =
+        a |> Foo.ofValue |> f
+
+
+let foo = FooBuilder ()
+
+let test =
+    foo {
+        apply (Foo.ofValue 5)
+        apply (Foo.ofValue true)
+        apply (Foo.ofValue 12.34)
+        apply (Foo.ofValue "Hello")
+        into (fun i b f s -> if i > 3 && b then s else "Nope")
+    }
+```
+
+Thanks to [nickcowle](https://github.com/nickcowle) for providing this example.
+
+</p></details>
+
+# Compatibility
+[compatibility]: #compatibility
+
+## Is this a breaking change?
+
+This change should be backwards compatible.
+
+Existing computation expression builders with an `Apply` method should not change in behaviour, since usages of the builder would still need to add the new `let! ... and! ...` syntax to activate it. In particular, in the case of `let! ... return ...`, we will continue to only pick bind, as mentioned earlier.
+
+## What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+Previous compiler versions reject the new `and!` and `anduse!` keywords:
+
+```
+error FS1141: Identifiers followed by '!' are reserved for future use
+```
+
+## What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+
+Since the syntax is desugared into a method call on the builder object, after compilation usages of this feature will be usable with previous compiler versions.
+
+# Unresolved Questions
+[unresolved]: #unresolved-questions
+
+Is `anduse!` an acceptable keyword for when `and!` must also imply a call to `Using`? Are there sensible alternatives?
```

Comment:
I'm guess I'm ok with `anduse!` all things considered, though perhaps we need a really compelling example where multiple resources are bound in a way that needs an imperative disposal.